### PR TITLE
Add rough support for extension operators

### PR DESCRIPTION
**Pull request by Schmavery**
Here's some rough work to see how we could support extension operators (##, #=) in reason.
This seems to work with bucklescript in our initial testing but we haven't managed to get everything together for js_of_ocaml (there seems to be some issue with <-). @chenglou 
Related: #650, #652

### CODE REVIEW

**Comment by facebook-github-bot-7** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by hhugo** [comment]
js_of_ocaml (ppx) syntax is `obj##meth`, `obj##.prop` and `obj##.prod := exp`

**Comment by yunxing** [comment]
@Schmavery What's the issue with "<-" ?

File: src/reason_lexer.mll [hunk 1]
```diff
@@ -446,6 +446,8 @@ rule token = parse
         lexbuf.lex_curr_p <- { curpos with pos_cnum = curpos.pos_cnum - 1 };
         STAR
       }
+  | "#" (appropriate_operator_suffix_chars | "#")
```
Comment (yunxing) [inline src/reason_lexer.mll]: This will only work with operators that are two characters long, right? I think you need to use a "?" after the parens.

Comment (Schmavery) [inline src/reason_lexer.mll]: I was actually just cleaning this up to address my previous issues with it and added a "+", which I believe is more appropriate?

Comment (yunxing) [inline src/reason_lexer.mll]: You are right, we should use "+" instead of "?".

**Comment by Schmavery** [comment]
It looks like the [jsoo code](https://github.com/ocsigen/js_of_ocaml/blob/master/examples/wiki/main.ml#L38) that we tried to use as a test example was actually using the old ocamlp4 proprocessor.  I added the '+' that I forgot on the lexer regex (based on [what ocaml does](https://github.com/ocaml/ocaml/blob/2c96a267dc3bf715d64aa134c031602d398d90ab/parsing/lexer.mll#L517)) and now I seem to be able to compile both jsoo and bucklescript examples (using the ppx syntax).

**Comment by chenglou** [comment]
cc @bobzhang, apparently this PR also works with BuckleScript!

**Comment by yunxing** [comment]
@Schmavery
Looks good to me. 

Maybe add some test cases with these operators? Especially how they interact with "<-" (you want to make sure it's left associative and has higher precedence than "<-")

File: src/reason_pprint_ast.ml [hunk 2]
```diff
@@ -405,6 +405,9 @@ let rules = [
       (fun s -> (`Right, s = "asr"));
     ];
     [
+      (fun s -> (`Left, s.[0] == '#'));
```
Comment (jordwalke) [inline src/reason_pprint_ast.ml]: Just to confirm - this order in the rules matches the order of token precedence defined in `reason_parser.mly`?

**Comment by jordwalke** [comment]
Looking great. Test cases would be very good. I'm currently rewriting all of that precedence checking/printing code (for infix combined with assignment `=`, `:=`, etc). But if the test cases are in place, it will ensure I don't break it!

**Comment by yunxing** [comment]
@jordwalke Good point about "<-".

@Schmavery :

`<-` is represented as `=` in Reason.
So in the example you were trying, it should be written as
```
preview##style##border = Js.string "1px black dashed";
```

**Comment by jordwalke** [comment]
So it looks great as long as the precedence is correct, and some test cases are added so that I make sure I don't break it in the diffs I'm about to send.

Some ideas for test cases:
- Making sure that it abides by the left precedence.  `x##y##z` should be parsed and printed exactly the same as `(x##y)##z` - both should be printed without parens. Likewise `x##(y##z)` should have its parens preserved when printing due to the left biased precedence.

**Comment by jaredly** [comment]
@jordwalke what could `x##(y##z)` possibly mean? surely not that `y##z` returns a string and then you access that property on `x`? I should think that property access would only be left-recursive

**Comment by bobzhang** [comment]
@jaredly x##method_name is still static, so in most cases, x##(y##z) is a compile error

**Comment by chenglou** [comment]
I'm literally doing `let result = js_method_run1 (!react)#createElement foo;` right now. Someone merge this before I rage merge this and break reason

**Comment by jordwalke** [comment]
This has merge conflicts. If someone resolves them, we can insta-merge as soon as the tests pass.

**Comment by chenglou** [comment]
It's just `reason_parser.messages`. I'll resolve locally and merge.

**Comment by jordwalke** [comment]
okay

**Comment by jordwalke** [comment]
You *coullldd* add a quick little test case too :D

**Comment by chenglou** [comment]
Ok, will add test when I wake up.

**Pull request merged**